# **Preparacion del ambiente**

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
#!pip install seaborn
import seaborn as sns
#!pip install plotly==5.7.0
import plotly.express as px
from numpy.ma.core import count

In [3]:
# agregamos código para que identifique más de un display en una celda
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

# **Importación de los datos de properati**


In [ ]:
#leo el archivo de properati
#en colab
csv_path ="/content/drive/MyDrive/DS - Curso/Clases/Desafio 1/Data/properatti.csv"
data = pd.read_csv(csv_path)

#en jupyter
#data = pd.read_csv("../../../data/properatti.csv", sep = ",", low_memory=False) 
#data.head(3)

## Completar info Habitaciones


Comprobamos el tipo y los valores.

In [ ]:
print("tipo  : ", data.rooms.dtype)
print("nulos : ", data.rooms.isnull().sum())
data.rooms.unique()

tipo  :  float64
nulos :  73830


array([nan,  1.,  4.,  3.,  2.,  6.,  5., 10.,  7.,  9.,  8., 17., 22.,
       15., 12., 11., 14., 16., 20., 13., 25., 19., 30., 18., 32., 24.,
       31., 21., 29., 27., 23., 28.])

In [ ]:
totalVentas = data.operation.notnull().sum()
totalSinCuartos = data.rooms.isnull().sum()
porcentajeSinCuartos = round( (totalVentas- totalSinCuartos) / totalVentas * 100, 2)
print('Total Registros \t\t\t', totalVentas )
print('Total sin ambientes \t\t\t', totalSinCuartos )
print('Porcentaje sin ambientes \t\t', porcentajeSinCuartos, ' %' )


Total Registros 			 121220
Total sin ambientes 			 73830
Porcentaje sin ambientes 		 39.09  %


El dataset esta completo en un 39.09% para todas las propiedades, podemos hacer foco en los departamentos:


A continuación se propone recuperar los datos faltantantes desde las columnas de título y descripción de la publicación. Ponemos en mayúscula los títulos y descripciones y hacemos una muestra de los que tienen incompletos `rooms`.

In [ ]:
data['title'] = data.title.str.upper()
data['description'] = data.description.str.upper()
data[['rooms', 'title', 'description']].sample(10)

rooms                                              title  \
19350     NaN  CASA 350M² CON 3 PLANTAS EN FERREYRA, ANDRES 4...   
70903     NaN                        DEPARTAMENTO EN P.CHACABUCO   
28809     3.0  EXCELENTE CONTRAFRENTE CON VISTA VERDE EN VENT...   
5899      4.0  DEPARTAMENTO A ESTRENAR DOS DORMITORIOS CON TE...   
63644     NaN  CASA ANTIGUA A RECICLAR DE 210 M2, SOBRE TERRE...   
59157     4.0             EXCELENTE 4 AMB.-EDIFICIO DE CATEGORIA   
98589     NaN           DEPARTAMENTO - BROWN AL 2700 - PICHINCHA   
85088     NaN                   EXCELENTE CASA DE LADRILLO VISTA   
111284    3.0              DUPLEX 3 AMBIENTES HAEDO APTO CREDITO   
86730     NaN      MONOAMBIENTE EN ALTURA APTO CREDITO-APTO PROF   

                                              description  
19350   DISTINTA A TODAS!!! DI VI NA CASA MODERNA ESTI...  
70903   AVISO LEGAL: LAS DESCRIPCIONES ARQUITECTÓNICAS...  
28809   EXCELENTE CONTRAFRENTE EN MUY BUEN EDIFICIO.LI...  
5899    DEPARTAMENTO DE 2 DORMITORIOS CON AMPLIO BALCÓ...  
63644   CHAJARI - CIUDAD DE TERMAS, EN LA RUTA DEL MER...  
59157   "VENTA EN COMUN CON GRUPOREAL S.A."-DPTO. 4 AM...  
98589   GARCIA ANDREU-NEGOCIOS INMOBILIARIOS- OFRECE E...  
85088   CASA CON 2 DORMITORIOS (1 EN PLANTA ALTA Y OTR...  
111284  CORREDOR RESPONSABLE: ELEONORA MEDEOT MINUJEN ...  
86730   ELEGANTE Y MODERNO EDIFICIO DE ASPECTO CLÁSICO...

Reemplazamos los números escritos como palabras y todas la variantes de llamar a los monoambientes por '1 AMBIENTE'.

In [ ]:
reemplazo = {"UNO":"1", "DOS":"2", "TRES":"3", "CUATRO":"4", "CINCO":"5", "SEIS":"6", "SIETE":"7", "OCHO":"8", "NUEVE":"9", "DIEZ":"10", "MONOAMBIENTE":"1 AMBIENTE", "UN":"1", "AMBIENTE DIVISIBLE":"1 AMBIENTE", "MONO AMBIENTE": "1 AMBIENTE"}

In [ ]:
for i in data.description:
  data.description[i].str.replace(reemplazo.keys, reemplazo.values, regex=False)

KeyError: ignored

In [ ]:
data.description = data.description.str.replace('UNO', '1', regex=False)
data.description = data.description.str.replace('DOS', '2', regex=False)
data.description = data.description.str.replace('TRES', '3', regex=False)
data.description = data.description.str.replace('CUATRO', '4', regex=False)
data.description = data.description.str.replace('CINCO', '5', regex=False)
data.description = data.description.str.replace('SEIS', '6', regex=False)
data.description = data.description.str.replace('SIETE', '7', regex=False)
data.description = data.description.str.replace('OCHO', '8', regex=False)
data.description = data.description.str.replace('NUEVE', '9', regex=False)
data.description = data.description.str.replace('DIEZ', '10', regex=False)
data.description = data.description.str.replace('MONOAMBIENTE', '1 AMBIENTE', regex=False)
data.description = data.description.str.replace('MONOAMB', '1 AMBIENTE', regex=False)
data.description = data.description.str.replace('UN', '1', regex=False)
data.description = data.description.str.replace('AMBIENTE DIVISIBLE', '1 AMBIENTE', regex=False)
data.description = data.description.str.replace('MONO AMBIENTE', '1 AMBIENTE', regex=False)

data.title = data.title.str.replace('UNO', '1', regex=False)
data.title = data.title.str.replace('DOS', '2', regex=False)
data.title = data.title.str.replace('TRES', '3', regex=False)
data.title = data.title.str.replace('CUATRO', '4', regex=False)
data.title = data.title.str.replace('CINCO', '5', regex=False)
data.title = data.title.str.replace('SEIS', '6', regex=False)
data.title = data.title.str.replace('SIETE', '7', regex=False)
data.title = data.title.str.replace('OCHO', '8', regex=False)
data.title = data.title.str.replace('NUEVE', '9', regex=False)
data.title = data.title.str.replace('DIEZ', '10', regex=False)
data.title = data.title.str.replace('MONOAMBIENTE', '1 AMBIENTE', regex=False)
data.title = data.title.str.replace('MONOAMB', '1 AMBIENTE', regex=False)
data.title = data.title.str.replace('UN', '1', regex=False)
data.title = data.title.str.replace('AMBIENTE DIVISIBLE', '1 AMBIENTE', regex=False)
data.title = data.title.str.replace('MONO AMBIENTE', '1 AMBIENTE', regex=False)


Recuperamos los ambientes desde título, descripción y actualizamos rooms.

In [ ]:
controlRooms = data[(data.rooms.isnull())]
controlRooms.rooms = controlRooms.description.str.extract("(\d+) AMB").astype(float)
data.update(controlRooms)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
controlRooms = data[(data.rooms.isnull()) ]
controlRooms.rooms = controlRooms.title.str.extract("(\d+) AMB").astype(float)
data.update(controlRooms)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
controlRooms = data[(data.rooms.isnull()) ]
controlRooms.rooms = controlRooms.description.str.extract("(\d+) DORM").astype(float) + 1
data.update(controlRooms)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
controlRooms = data[(data.rooms.isnull()) ]
controlRooms.rooms = controlRooms.title.str.extract("(\d+) DORM").astype(float) + 1
data.update(controlRooms)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
controlRooms = data[(data.rooms.isnull()) ]
controlRooms.rooms = controlRooms.title.str.extract("(\d+)AMB").astype(float)
data.update(controlRooms)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
controlRooms = data[(data.rooms.isnull()) ]
controlRooms.rooms = controlRooms.description.str.extract("(\d+)AMB").astype(float)
data.update(controlRooms)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
controlRooms = data[(data.rooms.isnull()) ]
controlRooms.rooms = controlRooms.title.str.extract("(\d+)DORM").astype(float) + 1
data.update(controlRooms)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
controlRooms = data[(data.rooms.isnull()) ]
controlRooms.rooms = controlRooms.description.str.extract("(\d+) DORM").astype(float) + 1
data.update(controlRooms)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
controlRooms = data[(data.rooms.isnull()) ]
controlRooms.rooms = controlRooms.title.str.extract("(\d+)HABITACIO").astype(float) + 1
data.update(controlRooms)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
controlRooms = data[(data.rooms.isnull()) ]
controlRooms.rooms = controlRooms.title.str.extract("(\d+) HABITACIO").astype(float) + 1
data.update(controlRooms)

In [ ]:
controlRooms = data[(data.rooms.isnull()) ]
controlRooms.rooms = controlRooms.description.str.extract("(\d+)HABITACIO").astype(float) + 1
data.update(controlRooms)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
controlRooms = data[(data.rooms.isnull()) ]
controlRooms.rooms = controlRooms.description.str.extract("(\d+) HABITACIO").astype(float) + 1
data.update(controlRooms)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
controlRooms = data[(data.rooms.isnull()) ].sample(10)
controlRooms[['rooms', 'title', 'description']]

rooms                                              title  \
108499    NaN                            DEPARTAMENTO - LA PLATA   
13777     NaN  VENTA LOCAL CON ENTREPISO APARTAMENTOS DEL CAMPUS   
117303    NaN                             DEPARTAMENTO - COGHLAN   
55836     NaN                              DEPARTAMENTO EN VENTA   
90402     NaN                              DEPARTAMENTO - CENTRO   
117731    NaN                              LOCAL - NUEVA CORDOBA   
106355    NaN                                  LOCAL - EL BOLSON   
105248    NaN                          CASA - CO1TRIES/B.CERRADO   
85437     NaN                               CASA EN MONTE GRANDE   
104527    NaN                                  LOCAL - EL BOLSON   

                                              description  
108499  AL FRENTE, PRIMER PISO POR ESCALERA, MUY BUEN ...  
13777   LOCAL A LA VENTA DENTRO DEL COMPLEJO APARTAMEN...  
117303  DEPARTAMENTO 40M² CON LUMINOSO EN MONROE AV. 3...  
55836   INSTALACIÓN ELECTRICA NUEVA, PREPARADA PARA 3 ...  
90402   DEPARTAMENTO CON ELECTRICIDAD, CÓRDOBA, CÓRDOB...  
117731  CARACTERISTICAS - LOCAL CON 19,40 M2 PROPIOS (...  
106355  FONDO DE COMERCIO DE FOTOGRAFÍA, REGALERIA Y E...  
105248  CASA 300M² CON ELECTRICIDAD, TUCUMÁN, YERBA BUENA  
85437   EXCELENTE PROPIEDAD EN MUY BUENA UBICACIÓN DES...  
104527  CONSULTORIO MEDICO DE 12,5M2,  N° 8 EN INSTITU...

Volvemos a controlar luego del proceso el grado de llenado de la columna

In [ ]:
totalVentas = data.operation.notnull().sum()
totalSinCuartos = data.rooms.isnull().sum()
porcentajeSinCuartos = round( (totalVentas- totalSinCuartos) / totalVentas * 100, 2)
print('Total Registros \t\t\t', totalVentas )
print('Total sin Ambientes \t\t\t', totalSinCuartos )
print('Porcentaje con ambientes \t\t', porcentajeSinCuartos, ' %' )

Total Registros 			 121220
Total sin Ambientes 			 16117
Porcentaje con ambientes 		 86.7  %


Luego del proceso se paso del 39,09 % de llenado para todas las propiedades del data set al 86,7 %. Veamos que pasa para cada tipo de propiedad

# COLUMNAS DUMMIES

### Creamos columnas con amenities adicionales que tendrán utilidad para predecir variación de precios.



In [ ]:
#fracciono la columna properti_url para sacar la nube de palabras mas repetidas
import re
patron_url = re.compile(pattern="_", flags =re.IGNORECASE)
lista_url = data["properati_url"].apply(lambda x : patron_url.split(x))
serie_palabras = pd.Series(np.hstack(lista_url))
#serie_palabras.value_counts().head(20).plot(kind="bar")


In [ ]:
serie_palabras.value_counts().head(50)

venta                            121086
departamento                      71019
garage                            56620
lavadero                          44034
balcon                            42286
casa                              40196
parrilla                          33381
piscina                           32146
luminoso                          31865
suite                             27101
placard                           26145
terraza                           25101
toilette                          22841
patio                             21168
vestidor                          18011
jardin                            17131
quincho                           12848
aire-acondicionado                12763
sum                               11665
dependencias                      11316
amenities                         10607
baulera                           10512
estrenar                          10379
vista                             10011
gimnasio                           9363


In [ ]:
# a partir de la nube de palabras selecciono las que son buenos adicionales
adicionales = ["garage", "balcon", "parrilla", "piscina", "terraza", "patio", "jardin", "quincho", "sum", "amenities", "baulera", "gimnasio", "subte-linea-d", "subte-linea-b", "subte-linea-a", "subte-linea-h", "subte-linea-e" ]


In [ ]:
#elimino el primer elemento de lista_url para no tener el elemento con el http: etc
for sublist in lista_url:
  del sublist[0]

In [ ]:
lista_url

0         [venta, ph, mataderos, lavadero, patio, inmobi...
1         [venta, departamentos, la-plata, balcon, lavad...
2         [venta, departamentos, mataderos, lavadero, pl...
3                   [venta, ph, liniers, patio, g-goffredo]
4         [venta, departamentos, centro, cristina-pavone...
                                ...                        
121215    [venta, departamento, belgrano, balcon, suite,...
121216    [venta, casa, beccar, suite, hidromasaje, jard...
121217    [venta, departamento, villa-urquiza, holmberg,...
121218    [venta, departamento, plaza-colon, lavadero, l...
121219    [venta, departamento, capital-federal, baulera...
Name: properati_url, Length: 121220, dtype: object

In [ ]:
#creo una función que compare la lista de palabras con la lista de listas
#y me da como resultado una lista de listas de palabras true/false segun coincida o no 
def buscador_palabras(quebuscar, dondebuscar):
  listadeextras = []
  for listas in dondebuscar:
    extras = []
    for palabra in quebuscar:
      if palabra in listas:
        extras.append(True)
      else:
        extras.append(False)
    listadeextras.append(extras)
  #print(listadeextras)
  return listadeextras     

In [ ]:
#aplico la funcion a mi lista "adicionales" y "lista_url"
#chequeo que tenga la misma longitud de data
resultado = buscador_palabras(adicionales,lista_url)
len(resultado)

121220

In [ ]:
#convierto resultado en dataframe, y renombro las columnas por la lista de palabras adicionales
df = pd.DataFrame(resultado)
df.columns = ["garage", "balcon", "parrilla", "piscina", "terraza", "patio", "jardin", "quincho", "s.u.m.", "amenities", "baulera", "gimnasio","subte-linea-d", "subte-linea-b", "subte-linea-a", "subte-linea-h", "subte-linea-e"]
df

garage  balcon  parrilla  piscina  terraza  patio  jardin  quincho  \
0        False   False     False    False    False   True   False    False   
1         True    True     False    False    False  False   False    False   
2        False   False     False    False    False  False   False    False   
3        False   False     False    False    False   True   False    False   
4        False   False     False    False    False  False   False    False   
...        ...     ...       ...      ...      ...    ...     ...      ...   
121215    True    True      True     True    False  False   False     True   
121216    True   False      True     True    False  False    True     True   
121217    True    True      True    False     True  False   False    False   
121218   False   False     False    False    False  False   False    False   
121219   False   False     False    False    False  False   False    False   

        s.u.m.  amenities  baulera  gimnasio  
0        False      False    False     False  
1        False      False    False     False  
2        False      False    False     False  
3        False      False    False     False  
4        False      False    False     False  
...        ...        ...      ...       ...  
121215   False      False    False     False  
121216   False      False    False     False  
121217   False       True    False     False  
121218   False      False    False     False  
121219   False      False     True     False  

[121220 rows x 12 columns]

In [ ]:
#uno el dataframe original con el nuevo generado de true/false
data = pd.merge(data,df,left_index=True, right_index=True)
data.columns
#antes me puso los dos indices como resultado del merge, y tuve que sacar la primera columna
#data.drop(columns=data.columns[0], axis=1,inplace=True)
#data.columns